core/src/main/java/org/infinispan/xsite/OfflineStatus.java

```diff
@@ -181,18 +219,60 @@ private boolean internalMinTimeHasElapsed() {
     */
    @GuardedBy("this")
    private boolean internalReset() {
-      boolean wasOffline = isOffline;
+      boolean wasOffline = isOffline(localStatus);
       firstFailureTime = NO_FAILURE;
       failureCount = 0;
-      isOffline = false;
       if (wasOffline) {
-         listener.siteOnline();
+         switchGlobally(localStatus);
       }
       return wasOffline;
    }
 
    @GuardedBy("this")
    private long internalMillisSinceFirstFailure() {
-      return timeService.timeDuration(MILLISECONDS.toNanos(firstFailureTime), MILLISECONDS);
+      return timeService.get().timeDuration(MILLISECONDS.toNanos(firstFailureTime), MILLISECONDS);
+   }
+
+   private void switchGlobally(long currentStatus) {
+      replaceInCache(currentStatus, currentStatus + 1, 0);
+      // optimistically switch assuming the “replace” eventually completes
+      updateLocalStatus(currentStatus + 1);
+   }
+
+   private void replaceInCache(long oldStatus, long newStatus, int retry) {
+      globalState.replaceAsync(key, oldStatus, newStatus).exceptionally(ignored -> {
+         if (retry <= MAX_RETRIES) {
+            replaceInCache(oldStatus, newStatus, retry + 1);
+         }
+         return Boolean.TRUE;
+      });
+   }
+
+   private void updateLocalStatus(long newStatus) {
+      synchronized (this) {
+         if (newStatus <= localStatus) {
+            return;
+         }
+         localStatus = newStatus;
+      }
+      if (isOnline(newStatus)) {
+         listener.siteOnline();
+      } else {
+         listener.siteOffline();
+      }
+   }
+
+   @Override
+   public synchronized boolean accept(Object key, Object oldValue, Metadata oldMetadata, Object newValue, Metadata newMetadata, EventType eventType) {
+      return Objects.equals(key, this.key) && newValue instanceof Long newStatus && newStatus > localStatus;
+   }
+
+   @CacheEntryCreated
+   @CacheEntryModified
+   public void onStatusChanged(CacheEntryEvent<ScopedState, Long> event) {
+      if (event.isPre() || event.getValue() == null || !key.equals(event.getKey())) {
```

Comment:
Isn't the `event.isPre()` check redundant when we have `Listener.Observation.POST`?

Reply:
All the conditions are redundant due to the cache filter. This won't be invoked often enough but I hope the C2 compiler removes this code.

Reply:
If they're redundant, why are we adding them?

Reply:
better safe than sorry. I converted them to `assert`